src/command/dev/next/router/run.rs

```diff
@@ -120,53 +127,141 @@ impl RunRouter<state::LoadRemoteConfig> {
 
-        let config_watcher = RouterConfigWatcher::new(FileWatcher::new(self.state.config_path));
+        let run_router_binary = RunRouterBinary::builder()
+            .router_binary(self.state.binary.clone())
+            .config_path(config_path.clone())
+            .supergraph_schema_path(schema_path.clone())
+            .and_remote_config(self.state.remote_config.clone())
+            .spawn(spawn)
+            .build();
+
+        let (_router_log_events, run_router_binary_subtask): (
+            UnboundedReceiverStream<Result<RouterLog, RunRouterBinaryError>>,
+            _,
+        ) = Subtask::new(run_router_binary);
+
+        let abort_router = SubtaskRunUnit::run(run_router_binary_subtask);
+
+        self.wait_for_healthy_router(&studio_client_config).await?;
+
+        Ok(RunRouter {
+            state: state::Watch {
+                abort_router,
+                config_path,
+                schema_path,
+            },
+        })
+    }
+
+    async fn wait_for_healthy_router(
+        self,
+        studio_client_config: &StudioClientConfig,
+    ) -> Result<(), RunRouterBinaryError> {
+        if !self.state.config.health_check_enabled() {
+            info!("Router healthcheck disabled in the router's configuration. The router might emit errors when starting up, potentially failing to start.");
+            return Ok(());
+        }
+
+        let healthcheck_endpoint = self.state.config.health_check_endpoint();
+
+        let healthcheck_client = studio_client_config.get_reqwest_client().map_err(|err| {
+            RunRouterBinaryError::Internal {
+                dependency: "Reqwest Client".to_string(),
+                error: format!("Failed to get client: {err}"),
+            }
+        })?;
+
+        let healthcheck_request = healthcheck_client
+            .get(healthcheck_endpoint.to_string())
+            .build()
+            .map_err(|err| RunRouterBinaryError::Internal {
+                dependency: "Reqwest Client".to_string(),
+                error: format!("Failed to build healthcheck request: {err}"),
+            })?;
+
+        let ten_secs = Duration::from_secs(10);
+        let start_time = Instant::now();
+        let mut success = false;
+
+        // Wait for the router to become healthy before continuing by checking its health endpoint,
+        // waiting only 10s
+        while !success && Instant::now().duration_since(start_time).le(&ten_secs) {
+            sleep(Duration::from_millis(100)).await;
+
+            let Some(request) = healthcheck_request.try_clone() else {
+                return Err(RunRouterBinaryError::Internal {
+                    dependency: "Reqwest Client".to_string(),
+                    error: "Failed to clone healthcheck request".to_string(),
+                });
+            };
+
+            if let Ok(res) = healthcheck_client.execute(request).await {
+                success = res.status().is_success()
+            }
+        }
```

Comment:
Could we do this a bit more cleanly with the timeout idiom from tokio? https://docs.rs/tokio/latest/tokio/time/fn.timeout.html

Reply:
yep! will make that change before merging